File: dbos_server_service/secret_service/src/api/v1/endpoints/dept_grants.py
"""HTTP-эндпоинты /credentials/{id}/dept-grants — только cross_department."""

from fastapi import APIRouter, Depends, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from src.core.config import get_settings
from src.core.limiter import limiter
from src.dependencies.auth import CurrentIdentity, require_user_context
from src.dependencies.db import get_db
from src.models import DeptGrant
from src.schemas.common import OkResponse
from src.schemas.dept_grants import DeptGrantCreate, DeptGrantList, DeptGrantRead
from src.services import dept_grant_service

router = APIRouter(prefix="/credentials", tags=["dept_grants"])


def _to_read(g: DeptGrant) -> DeptGrantRead:
    return DeptGrantRead(
        id=g.id,
        cred_id=g.cred_id,
        recipient_dept_id=g.recipient_dept_id,
        granted_by_user_id=g.granted_by_user_id,
        granted_at=g.granted_at,
    )


@router.post(
    "/{cred_id}/dept-grants",
    response_model=DeptGrantRead,
    status_code=status.HTTP_201_CREATED,
    summary="Выдать DeptGrant recipient-dep'у (owner dep_admin / service_admin)",
)
@limiter.limit(get_settings().rate_limit_dept_grant)
async def add_dept_grant(
    cred_id: str,
    request: Request,
    payload: DeptGrantCreate,
    identity: CurrentIdentity,
    db: AsyncSession = Depends(get_db),
) -> DeptGrantRead:
    require_user_context(identity)
    grant = await dept_grant_service.add(db, identity, cred_id, payload)
    return _to_read(grant)


@router.get(
    "/{cred_id}/dept-grants",
    response_model=DeptGrantList,
    summary="Список DeptGrant'ов для cross_department credential",
)
async def list_dept_grants(
    cred_id: str,
    identity: CurrentIdentity,
    db: AsyncSession = Depends(get_db),
) -> DeptGrantList:
    require_user_context(identity)
    grants = await dept_grant_service.list_for(db, identity, cred_id)
    return DeptGrantList(items=[_to_read(g) for g in grants])


@router.delete(
    "/{cred_id}/dept-grants/{grant_id}",
    response_model=OkResponse,
    summary="Снять DeptGrant (cascade RoleACL для recipient_dep)",
)
async def revoke_dept_grant(
    cred_id: str,
    grant_id: str,
    identity: CurrentIdentity,
    db: AsyncSession = Depends(get_db),
) -> OkResponse:
    require_user_context(identity)
    await dept_grant_service.revoke(db, identity, cred_id, grant_id)
    return OkResponse(ok=True)

File: dbos_server_service/web_ui/src/api/secret/deptGrants.ts
/**
 * Thin wrappers для `secret_service` `/secret/v1/credentials/{id}/dept-grants`.
 *
 * DeptGrant — только для cross_department кред: даёт recipient-dep'у право
 * получать RoleACL на креду. Снятие grant'а каскадит RoleACL recipient-dep'а.
 *
 * Source of truth:
 *   secret_service/src/api/v1/endpoints/dept_grants.py
 *   secret_service/src/schemas/dept_grants.py
 */

import { apiDelete, apiGet, apiPost } from "@/api/client";
import type {
  DeptGrant,
  DeptGrantCreateRequest,
  DeptGrantList,
  OkResponse,
} from "@/api/secret/types";

const BASE = "/secret/v1/credentials";

/** `GET /credentials/{id}/dept-grants` — список DeptGrant'ов кред. */
export function listDeptGrants(credId: string): Promise<DeptGrantList> {
  return apiGet<DeptGrantList>(`${BASE}/${credId}/dept-grants`);
}

/**
 * `POST /credentials/{id}/dept-grants` — выдать DeptGrant recipient-dep'у.
 *
 * `recipient_dept_id` обязателен. Гейтится owner dep_admin / admin
 * secret_service владеющего dep'а.
 */
export function addDeptGrant(
  credId: string,
  body: DeptGrantCreateRequest,
): Promise<DeptGrant> {
  return apiPost<DeptGrant>(`${BASE}/${credId}/dept-grants`, body);
}

/** `DELETE /credentials/{id}/dept-grants/{grantId}` — снять DeptGrant (cascade RoleACL). */
export function revokeDeptGrant(
  credId: string,
  grantId: string,
): Promise<OkResponse> {
  return apiDelete<OkResponse>(`${BASE}/${credId}/dept-grants/${grantId}`);
}
